### PR TITLE
LPIPS via torchmetrics, remove pyiqa op dep!

### LINKED ISSUE #1040: Remove pyiqa from LPIPS
pyiqa functionality should be removed from deepinv as per #891 . NIQE is taking some time to find a stable alternative #911. However, there are many well known implementations of LPIPS, and we should use one of these instead, deprecating pyiqa and testing the equivalence of the functionality offline.

e.g. https://lightning.ai/docs/torchmetrics/stable/image/learned_perceptual_image_patch_similarity.html https://github.com/richzhang/PerceptualSimilarity

### PR DESCRIPTION
**Pull request by Andrewwango**
This PR removes the pyiqa dependency from deepinv and closes #1040  #891 . For why we are removing pyiqa, see #891 .

- It changes LPIPS to use torchmetrics instead of pyiqa.
- It removes pyiqa as an optional dep. Note that NIQE still depends on it, so for now the NIQE test is skipped. The removal of pyiqa from NIQE is being done in #911 

To test the LPIPS in torchmetrics is identical to the previous, run this:

```python
import torch
butterflies = dinv.utils.load_example("butterfly.png", img_size=128).repeat(2, 1, 1, 1)
zeros = torch.zeros(2, 3, 128, 128)
rands = torch.rand(2, 3, 128, 128)

xs = [butterflies, zeros, rands, dinv.physics.PoissonNoise()(rands), dinv.physics.GaussianNoise()(butterflies), zeros + 0.1]
xs = [x.clip(min=0., max=1.) for x in xs]

old_lpips = dinv.metric.LPIPS()
new_lpips = LPIPS()

for x in xs:
    for y in xs:
        assert torch.allclose(old_lpips(y, x), new_lpips(y, x))
```

### Checks to be done before submitting your PR

- [x] `python3 -m pytest deepinv/tests` runs successfully.
- [x] `black .` runs successfully.
- [x] `make html` runs successfully (in the `docs/` directory).
- [x] Updated docstrings related to the changes (as applicable).
- [x] Added an entry to the [changelog.rst](https://github.com/deepinv/deepinv/blob/main/docs/source/changelog.rst).

### CODE REVIEW

**Comment by Andrewwango** [comment]
@nucli-vicky , do you mind reviewing this one? Thanks

**Comment by Andrewwango** [comment]
everything should be passing except the two examples which require NIQE

File: deepinv/loss/metric/perceptual.py [hunk 1]
```diff
@@ -33,23 +35,34 @@ class LPIPS(Metric):
...
+        if not (
+            torch.all((0.0 <= x_net) & (x_net <= 1.0))
+            and torch.all((0.0 <= x_net) & (x_net <= 1.0))
+        ):
+            raise ValueError("LPIPS metric requires x_net and x to be between 0 and 1.")
+
+        if not (x_net.shape[1] == x.shape[1] == 3):
+            raise ValueError(
+                f"LPIPS metric only supports 3-channel input, but got channels for x_net, x as {x_net.shape[1]}, {x.shape[1]}."
+            )
```
Comment (nucli-vicky) [inline deepinv/loss/metric/perceptual.py:63]: I'd additionally check if the tensors's shape length is four in case we get a 3D input (which is not supported by `torchmetrics` LPIPS)

File: docs/source/changelog.rst [hunk 2]
```diff
@@ -16,7 +16,8 @@ Changed
 - (Breaking) Make :class:`deepinv.physics.BlurFFT` compute a true convolution (now) instead of cross-correlation (before). It is now equivalent to :class:`deepinv.physics.Blur` with `padding="circular"` (:gh:`825` by `Minh Hai Nguyen`_). For even kernel sizes, the output is now shifted by one pixel to the top-left compared to before.
 - Refactor folder structure of least-squares solvers (:gh:`1011` by `Julian Tachella`_)
 - Removed `eps` parameter from :func:`deepinv.optim.linear.conjugate_gradient` (:gh:`1011` by `Julian Tachella`_)
-
+- :class:`deepinv.loss.metric.LPIPS` uses `torchmetrics` instead of `pyiqa`. (:gh:`1041` by `Andrew Wang`_)
```
Comment (nucli-vicky) [inline docs/source/changelog.rst:25]: Did we check if outputs are equivalent between both implementations? Otherwise we need to notify to users that results will be inconsistent with previous `deepinv` versions

File: deepinv/loss/metric/perceptual.py [hunk 3]
```diff
@@ -33,23 +35,34 @@ class LPIPS(Metric):
...
+    def __init__(self, device="cpu", **kwargs):
         super().__init__(**kwargs)
-        pyiqa = import_pyiqa()
-        self.lpips = pyiqa.create_metric(
-            "lpips", check_input_range=check_input_range, device=device, as_loss=as_loss
+        from torchmetrics.image.lpip import LearnedPerceptualImagePatchSimilarity
+
+        self.lpips = LearnedPerceptualImagePatchSimilarity(
+            net_type="alex", normalize=True, reduction="none"
         ).to(device)
```
Comment (nucli-vicky) [inline deepinv/loss/metric/perceptual.py]: I'd allow users to choose `net_type` and default to alexnet

File: deepinv/loss/metric/perceptual.py [hunk 4]
```diff
@@ -33,23 +35,34 @@ class LPIPS(Metric):
...
-        self.lower_better = self.lpips.lower_better
+        self.lower_better = True
 
     def metric(self, x_net, x, *args, **kwargs):
-        return self.lpips(x_net, x).squeeze(-1)
+        if not (
+            torch.all((0.0 <= x_net) & (x_net <= 1.0))
+            and torch.all((0.0 <= x_net) & (x_net <= 1.0))
```
Comment (mh-nguyen712) [inline deepinv/loss/metric/perceptual.py]: Should check both `x_net` and `x`, no?
We could do 
```
min, max = torch.aminmax(torch.cat([x_net, x], dim = 0))
valid = (min >= 0.0) & (max <=1)
``` 
it should be faster I think

File: deepinv/loss/metric/perceptual.py [hunk 5]
```diff
@@ -72,13 +85,15 @@ class NIQE(Metric):
...
     :Example:
 
-    >>> from deepinv.utils import load_example
-    >>> from deepinv.loss.metric import NIQE
-    >>> m = NIQE() # doctest: +IGNORE_RESULT
-    (...)
-    >>> x_net = load_example("celeba_example.jpg", img_size=128)
-    >>> m(x_net) # doctest: +ELLIPSIS
-    tensor([...])
+    ::
```
Comment (mh-nguyen712) [inline deepinv/loss/metric/perceptual.py:97]: Why remove the `>>>`? Do this still work for doctest?

Comment (Andrewwango) [reply deepinv/loss/metric/perceptual.py:97]: temporarily doing it to remove pyiqa from the doctest. But I will only merge this PR once #911 is merged anyway, so we can completely get rid of pyiqa

File: docs/source/changelog.rst (hunk 2 above)
Comment (Andrewwango) [reply docs/source/changelog.rst:25]: See my description (https://github.com/deepinv/deepinv/pull/1041#issue-3903777238): yes

Comment (Andrewwango) [reply docs/source/changelog.rst:25]: If you can think of a nicer way of automating this then it would be great! Could just include it as a redundant test, for those who are lucky enough to have pyiqa installed to run

File: deepinv/loss/metric/perceptual.py [hunk 6]
```diff
@@ -33,23 +35,34 @@ class LPIPS(Metric):
...
 
-    def __init__(self, device="cpu", check_input_range=False, as_loss=False, **kwargs):
+    def __init__(self, device="cpu", **kwargs):
         super().__init__(**kwargs)
-        pyiqa = import_pyiqa()
-        self.lpips = pyiqa.create_metric(
-            "lpips", check_input_range=check_input_range, device=device, as_loss=as_loss
+        from torchmetrics.image.lpip import LearnedPerceptualImagePatchSimilarity
+
+        self.lpips = LearnedPerceptualImagePatchSimilarity(
```
Comment (Andrewwango) [inline deepinv/loss/metric/perceptual.py]: - [x] TODO use functional interface instead (https://github.com/Lightning-AI/torchmetrics/issues/1224)

File: deepinv/loss/metric/perceptual.py (hunk 4 above)
Comment (Andrewwango) [reply deepinv/loss/metric/perceptual.py]: done

File: docs/source/changelog.rst (hunk 2 above)
Comment (nucli-vicky) [reply docs/source/changelog.rst:25]: Fair, missed that part of the description; my bad!

**Comment by Andrewwango** [comment]
TODO awaiting #911  merged , then the doc CI should be solved automatically

File: deepinv/loss/metric/perceptual.py [hunk 7]
```diff
@@ -21,35 +24,56 @@ class LPIPS(Metric):
...
+        if not (x_net.shape[1] == x.shape[1] == 3):
+            raise ValueError(
+                f"LPIPS metric only supports 3-channel input, but got channels for x_net, x as {x_net.shape[1]}, {x.shape[1]}."
+            )
+
+        min_val, max_val = torch.aminmax(torch.cat([x_net, x], dim=0))
+        if not ((min_val >= 0.0) & (max_val <= 1.0)):
+            raise ValueError("LPIPS metric requires x_net and x to be between 0 and 1.")
+
+        return self.lpips_fn(
```
Comment (mh-nguyen712) [inline deepinv/loss/metric/perceptual.py:69]: Can we load the net in __init__ to avoid loading the network at every call?

Comment (mh-nguyen712) [reply deepinv/loss/metric/perceptual.py:69]: Maybe use this? https://lightning.ai/docs/torchmetrics/stable/image/learned_perceptual_image_patch_similarity.html
